### PR TITLE
Unifiy Sarama config building

### LINKED ISSUE #268: Common util for SaramaConfig
**Problem**

the `distributed` channel has a good way to configure Sarama via values from a Configmap.

We want to move that to a common util, to be used in both: `distributed` and `consolidated`

It will provide a common way of handling (watching/parsing) of ConfigMap with full Sarama.Config. (and ability to merge in TLS/SASL info from `consolidated`)
common code for the sarama config

### PR DESCRIPTION
**Pull request by aliok**
Fixes #268

Follow up for https://github.com/knative-sandbox/eventing-kafka/pull/278

## Proposed Changes

- Move Sarama config building into `common` package
- Made the functions that modify the config more generic so that we can use them in both channels and the source
- Get rid of too specific functions for a module (like `UpdateSaramaConfig` function only made sense for the distributed channel but it could be replaced with a function that is more generic and reusable)
- Added more unit tests for the Sarama config building

**Release Note**

```release-note

```

**Docs**

### CODE REVIEW

**Comment by aliok** [comment]
Gonna add PR desc later. Let me see first if things are green.
Lacking some unit tests anyway.

File: cmd/channel/distributed/dispatcher/main.go
Comment (aliok) [inline cmd/channel/distributed/dispatcher/main.go:81]: This function was too specific. Got rid of it and changed its usages with using more generic `UpdateSaramaConfigWithKafkaAuthConfig`

File: cmd/channel/distributed/receiver/main.go
Comment (aliok) [inline cmd/channel/distributed/receiver/main.go:84]: Same story here

File: pkg/channel/consolidated/dispatcher/dispatcher.go
Comment (aliok) [inline pkg/channel/consolidated/dispatcher/dispatcher.go:100]: Once we decide to use the Sarama settings from `config-kafka`, we can simply pass it here (2nd arg).

File: pkg/channel/consolidated/utils/util.go
Comment (aliok) [inline pkg/channel/consolidated/utils/util.go:81]: Moved these into `common` to be able to reuse them in 2 channels and the source.

File: pkg/channel/distributed/common/kafka/sarama/sarama.go
Comment (aliok) [inline pkg/channel/distributed/common/kafka/sarama/sarama.go:53]: moved to `common`

File: pkg/channel/distributed/dispatcher/dispatcher/dispatcher.go
Comment (aliok) [inline pkg/channel/distributed/dispatcher/dispatcher/dispatcher.go:286]: If we have the Kafka auth in channel CRD for the distributed channel in the future, we can just pass it here.

**Comment by aliok** [comment]
@travis-minke-sap @eric-sap 
May I ask your eyes on this please?
Ready for review.

**Comment by travis-minke-sap** [comment]
/lgtm
/hold

* Holding in case @devguyio wants more time to review?

File: pkg/common/client/config.go
Comment (devguyio) [inline pkg/common/client/config.go:42]: I'd really avoid having functions with strong side effects like this. Again, unless I know  InitSaramaConfig internals, I'd be surprised that I'm passing an object to a function so that it sets a couple of default settings on the object I'm passing. I think we should avoid this style of functions with heavy side effects. 

alternatives I can think of is setting the default settings explicitly,  or my least favorite, call it something that describes its intent (e.g `UpdateConfigWithDefaults` or `SetConfigDefaults` )

File: pkg/common/client/config.go
Comment (devguyio) [inline pkg/common/client/config.go:84]: 1. I think the name isn't reflecting what's happening here as mentioned on the comment above. 
2. Would be good to have a comment on the function.

File: pkg/common/client/config.go
Comment (devguyio) [inline pkg/common/client/config.go:161]: It would be great to keep the type declaration at the top of the file

File: pkg/channel/consolidated/dispatcher/dispatcher.go
Comment (devguyio) [inline pkg/channel/consolidated/dispatcher/dispatcher.go:92]: I'm not sure I understand this comment

**Review by devguyio — CHANGES_REQUESTED** [review]
Nice work @aliok ! I can imagine it wasn't easy to figure out a unified approach. 

My 2 cents, I think your mind is in the right place, from what I can see you're trying to satisfy the different config needs of the different Kafka components but maybe we can take another stab at designing that config package and making sure it's well documented and easily understandable? cause it's used in three different components by different groups of maintainers. So check out my comments and lmk if they make sense. 

(e.g I can very clearly see that this is a typical builder pattern use case `...NewConfig().WithXYZ()` etc )

File: pkg/common/client/config.go
Comment (aliok) [reply pkg/common/client/config.go:32]: Hey, thanks for the review. You have a lot of good points.

It was kind of hard to refactor this. I initially when with a builder pattern locally, then saw it ended up too much duplicated code in the callers of the builder and changed my mind. I still think we shouldn't convert this into a builder.

To discuss individually:

1. This is about Sarama's, the Kafka client's, config. So, I named it as `common/client/config` but I don't have a strong opinion. I can move stuff around.
2. Ok, let me check that.
3. Gonna rework on the docs and also the naming of the functions.

File: pkg/common/client/config.go
Comment (aliok) [reply pkg/common/client/config.go:42]: I will go with `UpdateConfigWithDefaults` then.
Sounds good.

File: pkg/common/client/config.go
Comment (aliok) [reply pkg/common/client/config.go:84]: As above, I will rename this to `UpdateConfigWithDefaults` and add comments.

File: pkg/common/client/config.go
Comment (aliok) [reply pkg/common/client/config.go:161]: Sure

File: pkg/channel/consolidated/dispatcher/dispatcher.go
Comment (aliok) [reply pkg/channel/consolidated/dispatcher/dispatcher.go:92]: Let me add more explanation here.
But, it is about an explicit setting of the clientId here in the caller of the `BuildSaramaConfig`.

**Comment by aliok** [comment]
@devguyio switched to builder pattern. please have a new look

**Review by matzew — APPROVED** [review]
/lgtm

/hold

@devguyio please take a look and unhold if your comments are all addressed

File: pkg/common/client/config.go
Comment (devguyio) [inline pkg/common/client/config.go:92]: is there a reason of using a pointer here?

File: pkg/common/client/config.go
Comment (devguyio) [inline pkg/common/client/config.go:94]: same pointer question

File: pkg/common/client/config.go
Comment (devguyio) [inline pkg/common/client/config.go:97]: same pointer question

File: pkg/common/client/yaml.go
Comment (devguyio) [inline pkg/common/client/yaml.go:42]: why return a mix of value and pointer? probably the pointer isn't needed

**Review by devguyio — CHANGES_REQUESTED** [review]
Great job! I like the design! Just some cleanup regarding pointer-value semantics and it'd be ready to go!

**Review by devguyio — APPROVED** [review]
/lgtm
/unhold

**Pull request merged**
